File: LeetCode/Hao/49GroupAnagrams.java
public class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> hashMap = new HashMap<String, List<String>>();
        
        Arrays.sort(strs);
        
        for (int i = 0; i < strs.length; i ++) {
            char[] c = strs[i].toCharArray();
            Arrays.sort(c);
            String tmp = new String(c);
            
            if (hashMap.containsKey(tmp) == true) {
                hashMap.get(tmp).add(strs[i]);
            } else {
                List<String> l = new ArrayList<String> ();
                l.add(strs[i]);
                hashMap.put(tmp, l);
            }
        }
        List<List<String>> result = new ArrayList<List<String>> ();
        
        for (Map.Entry<String, List<String>> entry: hashMap.entrySet()) {
            result.add(entry.getValue());
        }
        
        return result;
    }
}

File: LeetCode/Peng/49GroupAnagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        int n = strs.size();
        vector<vector<string>> result;
        unordered_map<string, vector<string>> category;
        for (int i = 0; i < n; i++) {
            string temp = strs.at(i);
            sort(temp.begin(), temp.end());
            category[temp].push_back(strs.at(i));
        }
        unordered_map<string, vector<string>>::iterator iter;
        for (iter = category.begin(); iter != category.end(); iter++) {
            sort(iter->second.begin(), iter->second.end());
            result.push_back(iter->second);
        }
        return result;
    }
};